eras/alonzo/impl/testlib/Test/Cardano/Ledger/Alonzo/Binary/CostModelsSpec.hs

```diff
@@ -0,0 +1,183 @@
+  BSL.ByteString ->
+  IO ()
+checkValidCostModelDecoder _ cmsBytes ppuBytes = do
+  let ver = (eraProtVerHigh @era)
+  let cmDecoded = decodeFull @CostModels ver cmsBytes
+  cmDecoded `shouldBeRightAndSatisfy` validCm
+
+  let ppuDecoded = decodeFull @(PParamsUpdate era) ver ppuBytes
+  ppuDecoded
+    `shouldBeRightAndSatisfy` ( \ppu ->
+                                  (validCm <$> ppu ^. ppuCostModelsL) == SJust True
+                              )
+  where
+    validCm (CostModels valid errors unknown) =
+      (not . null) valid && null errors && null unknown
+
+checkInvalidCostModelDecoder ::
+  forall era.
+  AlonzoEraPParams era =>
+  Proxy era ->
+  BSL.ByteString ->
+  BSL.ByteString ->
+  IO ()
+checkInvalidCostModelDecoder _ cmsBytes ppuBytes = do
+  let ver = (eraProtVerHigh @era)
+  let cmDecoded = decodeFull @CostModels ver cmsBytes
+  let ppuDecoded = decodeFull @(PParamsUpdate era) ver ppuBytes
+  -- pre-Conway we are failing when deserializing invalid costmodels
+  if (ver < natVersion @9)
+    then
+      cmDecoded
+        `shouldBeLeftAndSatisfy` ( \case
+                                    DecoderErrorDeserialiseFailure txt _ -> txt == "CostModels"
+                                    _ -> False
+                                 )
+    else -- post-Conway, we are collecting CostModels deserialization errors
+
+      cmDecoded `shouldBeRightAndSatisfy` invalidCm
+
+  -- in no era are we deserializing invalid costmodels within PParamsUpdate
+  ppuDecoded
+    `shouldBeLeftAndSatisfy` ( \case
+                                DecoderErrorDeserialiseFailure _ _ -> True
+                                _ -> False
+                             )
+  where
+    invalidCm (CostModels valid errors _) =
+      null valid && (not . null) errors
+
+genValidCostModelsEnc ::
+  Language ->
+  Gen Encoding
+genValidCostModelsEnc lang =
+  genCostModelsEncoding lang (costModelParamsCount lang)
+
+genInvalidCostModelsEnc ::
+  Language ->
+  Gen Encoding
+genInvalidCostModelsEnc lang = do
+  let validCount = costModelParamsCount lang
+  count <- choose (0, validCount - 1)
+  genCostModelsEncoding lang count
+
+genCostModelsEncoding :: Language -> Int -> Gen Encoding
+genCostModelsEncoding lang count = (fromPlainEncoding . toCBOR) <$> enc
+  where
+    enc :: Gen Enc.Enc
+    enc = do
+      values <- vectorOf count (arbitrary :: Gen Int)
+      pure $
+        mconcat
+          [ Enc.E (TkMapLen 1)
+          , Enc.E (TkInt (fromEnum lang))
+          , Enc.E TkListBegin
+          , Enc.Em (Enc.E . TkInt <$> values)
+          , Enc.E TkBreak
+          ]
+
+genPParamsUpdateEnc ::
+  forall era.
+  Era era =>
+  Proxy era ->
+  Encoding ->
+  Gen Encoding
+genPParamsUpdateEnc _ cmEnc = (fromPlainEncoding . toCBOR) <$> enc
+  where
+    enc :: Gen Enc.Enc
+    enc = do
+      pure $
+        mconcat
+          [ Enc.E (TkMapLen 1)
+          , Enc.E (TkInt 18)
+          , Enc.E (toPlainEncoding (eraProtVerHigh @era) cmEnc)
+          ]
```

Comment:
This is not a `Gen`, it has no randomness. Also usage of `E` is unnecessary:
```suggestion
genPParamsUpdateEnc ::
  forall era.
  Era era =>
  Proxy era ->
  Encoding ->
  Encoding
genPParamsUpdateEnc _ = 
  fromPlainEncoding . toCBOR . Map.singleton (18 :: Int) . toPlainEncoding (eraProtVerHigh @era)
```

Reply:
:exploding_head: